### PR TITLE
feat(helm): adding-cpu-memory-resources

### PR DESCRIPTION
**Pull request by gardlt**
## Description

Adding Resource limitation on the deployments of the faas-netes

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

I have deployed the helm chart and the kubernetes manifest in a minikube instance. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Looks reasonable. I'd like to get a couple of people to look at the options.

**Comment by rawkode** [comment]
Looks decent. Has there been any profiling of faas-netes to ensure the defaults we provide are mostly sufficient for common use-cases?

**Comment by alexellis** [comment]
No, that is my concern by imposing a "finger in the air" limit.

**Comment by weikinhuang** [comment]
This is definitely a good idea. It's usually best practice to have resource constraints so the scheduler doesn't create hotspots.

**Comment by gardlt** [comment]
So the beauty of the helm chart is that you will be able to enable/disable the resources feature. Removing the "finger in the air" limit, I believe. For the hardcoded manifest, if we don't want it to impose we could remove...

**Comment by weikinhuang** [comment]
If we can get average usage data for all the control plane components, we can easily add default resource requests and limits, and a user would be able to override it as they please.

**Comment by weikinhuang** [comment]
Can we also instead change the format to something similar to the ones in the charts repository. Ex. https://github.com/kubernetes/charts/blob/master/stable/grafana/templates/deployment.yaml#L63

File: openfaas/templates/faas.yaml [hunk 1]
```diff
@@ -72,6 +72,10 @@ spec:
         image: {{ .Values.images.controller }}
         {{- end }}
         imagePullPolicy: Always
+        {{ if .Values.resources.enabled }}
+        resources:
+{{ toYaml .Values.resources.faasnetesd | indent 12 }}
```
Comment (alexellis) [inline openfaas/templates/faas.yaml]: @weikinhuang any thoughts for review?

**Comment by alexellis** [comment]
The limits do seem high. The University of Calgary are running the swarm versions with 20mb of RAM as a limit for the gateway etc. @weikinhuang what is the best way to profile this? Would we get a good idea by running some kind of balanced load-test?

Can you guys work together on this via Slack? @weikinhuang @gardlt @rawkode we have the #faas-netes channel on Slack. :+1:

**Comment by alexellis** [comment]
@gardlt

**Comment by rawkode** [comment]
Based on my findings, I'd say 32MiB RAM and 50m CPU

Anyone going above that is probably going to be tweaking their values anyway.

**Comment by alexellis** [comment]
I'm not sure if @gardlt is still working on this?

We now have PR https://github.com/openfaas/faas-netes/pull/73 from @resouer also

We need to add limits to force the correct QoS policy for the system-level pods.

**Comment by alexellis** [comment]
@rawkode would you like to take over from @gardlt .. this PR seems to have been abandoned?

**Comment by rawkode** [comment]
@alexellis Looks like you merged #37 instead?

**Comment by alexellis** [comment]
This is #37 @rawkode

**Comment by rawkode** [comment]
I meant #73 :joy:

File: openfaas/values.yaml [hunk 2]
```diff
@@ -50,3 +50,47 @@ ingress:
...
+  queueworker:
+    requests:
+      memory: "32MiB"
+      cpu: "50m"
+    limits:
+      memory: "32MiB"
+      cpu: "50m"
+  prometheus:
+    requests:
+      memory: "32MiB"
```
Comment (alexellis) [inline openfaas/values.yaml]: Needs 500MB RAM according to @stefanprodan.

Comment (stefanprodan) [reply openfaas/values.yaml]: This is the RAM usage after 12h on GKE:
openfaas

**Pull request closed without merging**